apps/hash-frontend/src/pages/new/types/select-generated-property-types.tsx

```diff
@@ -0,0 +1,236 @@
+import { PropertyType } from "@blockprotocol/type-system";
+import { PropertyTypeWithMetadata, Subgraph } from "@local/hash-subgraph/.";
+import { getPropertyTypesByBaseUrl } from "@local/hash-subgraph/stdlib";
+import {
+  Checkbox,
+  CircularProgress,
+  Collapse,
+  FormControl,
+  FormControlLabel,
+  formControlLabelClasses,
+  FormGroup,
+  FormLabel,
+  Typography,
+} from "@mui/material";
+import { Box } from "@mui/system";
+import { debounce } from "lodash";
+import {
+  FunctionComponent,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+import { useBlockProtocolQueryPropertyTypes } from "../../../components/hooks/block-protocol-functions/ontology/use-block-protocol-query-property-types";
+import { useAgentRunner } from "../../../components/hooks/use-agent-runner";
+import { useGenerateTypeUrlsForUser } from "../../shared/use-generate-type-urls-for-user";
+
+export type PropertyTypeDefinition = {
+  title: string;
+  description: string;
+  dataType: "text" | "number" | "boolean";
+};
+
+type SelectGeneratedPropertyTypesProps = {
+  entityTypeTitle: string;
+  entityTypeDescription: string;
+  setSelectedPropertyDefinitions: (
+    propertyDefinitions: (PropertyTypeDefinition | PropertyType)[],
+  ) => void;
+};
+
+export const SelectGeneratedPropertyTypes: FunctionComponent<
+  SelectGeneratedPropertyTypesProps
+> = ({
+  entityTypeTitle,
+  entityTypeDescription,
+  setSelectedPropertyDefinitions,
+}) => {
+  const { queryPropertyTypes } = useBlockProtocolQueryPropertyTypes();
+  const generateTypeUrlsForUser = useGenerateTypeUrlsForUser();
+
+  const [generatePropertyTypes, { loading }] = useAgentRunner(
+    "generate-entity-type-property-types",
+  );
+
+  const [
+    generatedPropertyTypeDefinitions,
+    setGeneratedPropertyTypeDefinitions,
+  ] = useState<(PropertyTypeDefinition | PropertyType)[]>();
+  const [selectedPropertyTypeTitles, setSelectedPropertyTypeTitles] = useState<
+    string[]
+  >([]);
+
+  const [allPropertyTypes, setAllPropertyTypes] = useState<Subgraph>();
+
+  const fetchAllPropertyTypes = useCallback(async () => {
```

Comment:
Because this is an `async` function, I'm not sure I agree it simplifies things by a lot as to my understanding it would require wrapping the implementation of the method in an anonymous function like so:
```TypeScript
  useEffect(() => {
    void (async () => {
      if (!allPropertyTypes) {
        const { data: propertyTypesSubgraph } = await queryPropertyTypes({
          data: {},
        });

        if (!propertyTypesSubgraph) {
          throw new Error(
            "Could not query property types to get all property types.",
          );
        }

        setAllPropertyTypes(propertyTypesSubgraph);
      }
    })();
  }, [allPropertyTypes, fetchAllPropertyTypes, queryPropertyTypes]);
```
And in general this is something that I like to avoid for readability purposes. But if you feel strongly about this, or there's something I'm missing about your suggestion, I'm happy to address.

Reply:
Defining the function within the effect means one the dependencies of the effect are defined where the effect is defined, instead of in a useCallback for the function – so there's a lot more clarity. It also makes it more obvious if there needs to be any clean up logic. 

You can either do what you've done here – or just assign it to a variable inside the effect and call with void.

Reply:
Alright, to be honest I'm not entirely convinced by this but don't feel strongly so have refactored in https://github.com/hashintel/hash/pull/2486/commits/7cfc6de80a3a5d14c480f4043b69ffc306555d73